doc/threshold/SONiC Threshold feature spec.md

```diff
@@ -0,0 +1,589 @@
+
+This command is used to configure a threshold for a specific unicast/multicast queue of a port. The threshold value is provided in %. Valid values are 1-100.
+
+### 3.6.3 Show Commands
+1) show priority-group threshold {shared \| headroom}
+
+This show command shows the threshold configuration for the shared/headroom priority-group buffers of all ports.
+
+2) show queue threshold {unicast \| multicast}
+
+This show command shows the threshold configuration for the unicast/multicast queue buffers of all ports.
+
+3) show threshold breaches
+
+This command shows the threshold breaches recorded by the system. 
+
+### 3.6.4 Clear commands
+
+1) clear priority-group threshold {port_alias} {PGindex} {shared \| headroom}
+
+This command can be used to clear a previously configured threshold on shared/headroom priority-group buffer of a port.
+
+2) clear priority-group threshold 
+
+This command can be used to clear all currently configured priority-group thresholds. 
+	
+3) clear queue threshold {port_alias} {queueindex} {unicast \| multicast}
+
+This command can be used to clear a previously configured threshold on unicast/multicast queue buffer of a port.
+
+4) clear queue threshold 
+
+This command can be used to clear all currently configured priority-group thresholds.
+
+5) clear threshold breach {eventid}
+
+This command can be used to clear all/particular threshold breaches recorded by the system. The event-id when provided specifies the breach event index (index is indicated in the output of show threshold breaches).
+
+### 3.6.5 Debug Commands
+Debug commands will be available once the debug framework is available. The debug commands are needed to dump:
+
+- Threshold entries and TAM object bindings maintained by ThresholdOrch.
+- Incoming protobuf data at ThresholdMgr.
+
+### 3.6.6 REST API Support
+N/A
+
+# 4 Flow Diagrams
+## 4.1 Config call flow 
+
+![Threshold config call flow](images/Threshold_call_flow.jpg)
+
+## 4.2 Breach event call flow
+
+![Threshold breach event call flow](images/Threshold_breach_call_flow.jpg)
+
+# 5 Error Handling
+
+## CLI
+- CLI configuration sanity is enforced by the CLI handler and any invalid configuration is rejected. An error is displayed to the user notifying the reason for rejection of the configuration.
+
+## ThresholdMgr
+- Any errors encountered during parsing of protobuf message are logged.
+
+## ThresholdOrch
+- Any error occurring in the orchestration agent is logged appropriately via SWSS logging.
+- Errors or failures of SAI APIs are logged by ThresholdOrch.
+- On failure of a SAI TAM API in the config sequence of section 3.5, the previously configured steps are rolled back i.e previously created intermediate TAM objects for threshold, event etc are destroyed. 
+
+# 6 Serviceability and Debug
+Debug commands specified in section 3.6.5 will be supported once the debug framework is available. 
+
+# 7 Warm Boot Support
+Warm boot is supported for the threshold feature.
+
```

Comment:
How are the notifications handled during warmboot? Does orchagent disable the thresholds during warm-shutdown, and enable them back during warm-boot?

Reply:
Any breach notifications received during warm-reboot will be lost.